modules/js-and-css/optimize-oembeds/hooks.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * Hook callbacks used for oEmbed Optimizer.
+ *
+ * @since n.e.x.t
+ * @package performance-lab
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+$oembed_lazy_load_scripts = false;
+
+/**
+ * Filter the oEmbed HTML.
+ *
+ * Add loading="lazy" to any iframe tags.
+ * Lazy load any script tags.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $html The oEmbed HTML.
+ * @return string
+ */
+function perflab_optimize_oembed_html( $html ) {
+	global $oembed_lazy_load_scripts;
+
+	// Find iframes and script tags in the oEmbed using the WP_HTML_Tag_Processor.
+	$p = new WP_HTML_Tag_Processor( $html );
+	if ( $p->next_tag( 'iframe' ) ) {
+		if ( empty( $p->get_attribute( 'loading' ) ) ) {
+			$p->set_attribute( 'loading', 'lazy' );
+			$html = $p->get_updated_html();
+		}
+	}
+	$p = new WP_HTML_Tag_Processor( $html );
+	if ( $p->next_tag( 'script' ) ) {
+		$oembed_lazy_load_scripts = true;
+		$p->set_attribute( 'data-lazy-embed-src', $p->get_attribute( 'src' ) );
+		$p->set_attribute( 'src', '' );
+		$html = $p->get_updated_html();
+	}
```

Comment:
to avoid scanning the document twice we can scan `next_tag()` and attach our behaviors based on the currently-matched tag name. the Tag Processor has to scan every tag anyway, so we're not skipping anything other than an extra function call per tag.

```php
$p = new WP_HTML_Tag_Processor( $html );

// These are only important if we expect to encounter multiple scripts
// or iframes but only want to process the first. This is here to preserve
// the behavior of the snippet in the PR, but I don't know if that was
// intentional or happenstance.
$have_seen_iframe = false;
$have_seen_script = false;

while ( $p->next_tag() ) {
	if ( 'IFRAME' === $p->get_tag() && ! $has_seen_iframe ) {
		$has_seen_iframe = true;

		$loading_value = $p->get_attribute( 'loading' );
		if ( ! is_string( $loading_value ) || empty( $loading_value ) ) {
			$p->set_attribute( 'loading', 'lazy' );
		}
	} else if ( 'SCRIPT' === $p->get_tag() && ! $has_seen_script ) {
		$has_seen_script = true;

		$oembed_lazy_load_scripts = true;
		$p->set_attribute( 'data-lazy-embed-src', $p->get_attribute( 'src' ) );
		$p->set_attribute( 'src', '' );
	}

	if ( $has_seen_iframe && $has_seen_script ) {
		break;
	}
}

return $p->get_updated_html();
```

Reply:
> the Tag Processor has to scan every tag anyway, so we're not skipping anything other than an extra function call per tag.

Ah, that is good to know, I will adjust accordingly. in general these HTML blocks will have either an iframe or a script tag, so definitely makes sense to only make one pass! Thanks.

Reply:
> ! is_string( $loading_value )

@dmsnell why is this check needed (in `if ( ! is_string( $loading_value ) || empty( $loading_value ) ) {`)

Reply:
if it's no problem to set them all as lazy it gets simpler in the code.

```php
$p = new WP_HTML_Tag_Processor( $html );
while ( $p->next_tag() ) {
	switch ( $p->get_tag() ) {
		case 'IFRAME':
			$loading_value = $p->get_attribute( 'loading' );
			if ( ! is_string( $loading_value ) || empty( $loading_value ) ) {
				$p->set_attribute( 'loading', 'lazy' );
			}
			break;

		case 'SCRIPT':
			$oembed_lazy_load_scripts = true;
			$p->set_attribute( 'data-lazy-embed-src', $p->get_attribute( 'src' ) );
			$p->set_attribute( 'src', '' );
			break;
	}
}

return $p->get_updated_html();
```

Reply:
removed is_string check in https://github.com/WordPress/performance/pull/886/commits/3701f2d04048ac4445bdbb3aab134f2c45a50732 - since `get_tag` wither returns a string or null and `empty` works as expected on either

Reply:
@adamsilverstein if an attribute is a boolean attribute then in the DOM it will return an empty string but the HTML API expresses that as `true` so that code can recognize the difference. unfortunately, `empty(true) === false`, so we might think there's a value set when in fact it's empty in the DOM

Reply:
(the only practical drawback here is that we'll miss adding the lazy attribute if we encounter something like `<script loading src="…">`, which may not seem like a big deal because it's still safe, just not 100% correct)